department-of-veterans-affairs/caseflow
# Check if the user is the judge or attorney on a case review to determine what actions they see in LegacyTasks

Resolves issue is some cases where a legacy appeal is sent back to the decision drafting attorney but they only see judge actions as they are an avlj.

### Description
Another hacky fix in the ever evolving saga of "how the heck do we determine if a legacy case is assigned to an acting judge to write a decision to to sign a decision". Some context is provided in #14886.

Initially, we tried to use the logic that if a decision had already been drafted for an appeal, this appeal is probably assigned to the user to sign the decision as a judge.

To implement this, we initially checked if an attorney case review existed for the case in caseflow. While this was an alright start, sometimes cases are sent from an attorney to a judge outside of caseflow and no attorney case review will exist.

To resolve this, we moved to checking to see if there was a decision document listed on the vacols case assignment. The thought process here is that this will always be populated whether sent to the judge in caseflow or vacols.

The problem with this is outlined in #14886. If the case returns to the acting judge, say for rewriting the decision, our code assumes they are being assigned the case to sign the decision they wrote.

For some of these cases that _were_ worked in caseflow, we can reach a happier middleground. We should check to see if they are the attorney that created the case review. If they are, the task will be an attorney task. It they were the judge on the case review, the task will be a judge task. If neither, we can fall back to old logic of checking to see if there is a decision already written.

While this will not resolve all cases, it will cut down on the number of batteam requests we get about this.

For instance, [this batteam request](https://dsva.slack.com/archives/CHX8FMP28/p1604942204005500) could have been avoided if we had implemented this check.
```ruby
user = User.find_by(css_id: "bvamsopko".upcase)
user.acting_judge_in_vacols?
=> true

appeal = LegacyAppeal.find_by(vacols_id: "3848262")
appeal.vacols_case_review.valid_document_id?
=> true
# based on our current logic, we should show this user judge actions as a decision exists already. SAD

appeal.attorney_case_review.attorney == user
=> true
# !! information in caseflow exists that definitively stipulates this user was the attorney on the case, we should show them attorney actions 
```

### Acceptance Criteria
- [ ] If an acting judge was the drafting attorney on a legacy case in caseflow, they should see attorney actions for the case assigned to them
- [ ] If an acting judge was the reviewing judge on a legacy case in caseflow, they should see judge actions for the case assigned to them
- [ ] If there is no attorney case review or if the user was neither the judge or attorney, fall back to checking if a decision has already been written for the case.

### Testing plan
1. Ensure das deprecation is turned off
```ruby
FeatureToggle.enabled?(:legacy_das_deprication)
=> false
```
2. Log in as bvaaaaaaabshire and go to your assign queue
1. Assign a case to an acting judge
1. If you get a "Case already assigned" error, you may need to fix some vacols data
```ruby
judge = User.find_by(css_id: "BVAAABSHIRE")
judge.vacols_uniq_id
=> "ID6"
# Should be "AABSHIRE"

VACOLS::Staff.where(sdomainid: "BVAAABSHIRE").pluck(:slogid)
=> ["ID6", "ID7", "AABSHIRE"]
VACOLS::Staff.where.not(slogid: "AABSHIRE").where(sdomainid: "BVAAABSHIRE").destroy_all
VACOLS::Staff.where(sdomainid: "BVAAABSHIRE").pluck(:slogid)
=> ["AABSHIRE"]

# Clear our cache and try again 
Rails.cache.delete("#{Rails.env}_staff_record_#{judge.css_id}")
User.find_by(css_id: "BVAAABSHIRE").vacols_uniq_id
=> "AABSHIRE"
```
5. If you get a "Deteam cannot be blank" error
```ruby
acting_judge = User.find_by(css_id: "BVAACTING")
acting_judge.vacols_group_id
=> ""

VACOLS::Staff.find_by(sdomainid: "BVAACTING").update!(stitle: "D#{Random.rand(1..5)}")
Rails.cache.delete("#{Rails.env}_staff_record_#{acting_judge.css_id}")
User.find_by(css_id: "BVAACTING").vacols_group_id
=> "D5"
```
6. Log in as the acting judge and go through checkout to send the case back to the judge
1. Cases are being sent back to these avljs through vacols, so let's do that manually
```ruby
VACOLS::Case.find(3662856).update_vacols_location!(acting_judge.vacols_uniq_id)
```
8. Return to the case and ensure the user has "Decision ready for review" action, not "Ready for dispatch"
1. Checkout master to see what the user used to see

### UI changes

Before (user sees judge actions)|After (user sees attorney actions)
---|---
<img width="728" alt="Screen Shot 2020-11-09 at 5 47 10 PM" src="https://user-images.githubusercontent.com/45575454/98605594-ae147a00-22b3-11eb-9814-e3fe696f630b.png">|<img width="728" alt="Screen Shot 2020-11-09 at 5 46 50 PM" src="https://user-images.githubusercontent.com/45575454/98605596-ae147a00-22b3-11eb-9c9e-446420c745f0.png">
